### ISSUE TITLE
Report type checks which always fail in analyzer

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#46170]
For example:

```dart
double foo() => 1;

main() {
  print(foo() as int);
}
```

With null safety this code always throws an error and dartanalyzer can figure this out using static types and issue a hint or lint.
Such type checks happened in practice in package:flutter_quill (fix: https://github.com/singerdmx/flutter-quill/pull/239/files).

**Comment by srawlins** [comment]
@alexmarkov what check are you looking for? Given:

```dart
class A {}
class B {}

...

B f(A a) => a as B;
```

There can be classes which extend both A and B.

Are you just thinking of the un-extendable classes (`int`, `double`, `num`, `String`, `bool`, `Future`, `FutureOr`, `void`, `Never`, or something like this)? Given the prevalence of int, double, String, etc, I think this could be useful.

**Comment by srawlins** [comment]
Yeah, whole program will be a neat feature in the future. Analyzer does not do it today.

I think sealed built-in classes will be most useful. Private classes with no public subtypes is not as useful.
